File: vm/java/top.java
/*
 * Copyright (c) 2009, Simon Waite and Charles Childers
 * Based on the C# implementation
*/

import java.*;
import java.io.*;


/**
 * @author Simon Waite
 * @author Charles Childers
 *
 * The retro class contains the Ngaro VM and an initial
 * image file for the Retro language.
 */
public class retro {
  int sp, rsp, ip;
  int data[], address[], ports[];
  int memory[];

  public static final int VM_NOP = 0;       public static final int VM_LIT = 1;       public static final int VM_DUP = 2;
  public static final int VM_DROP = 3;      public static final int VM_SWAP = 4;      public static final int VM_PUSH = 5;
  public static final int VM_POP = 6;       public static final int VM_CALL = 7;      public static final int VM_JUMP = 8;
  public static final int VM_RETURN = 9;    public static final int VM_GT_JUMP = 10;  public static final int VM_LT_JUMP = 11;
  public static final int VM_NE_JUMP = 12;  public static final int VM_EQ_JUMP = 13;  public static final int VM_FETCH = 14;
  public static final int VM_STORE = 15;    public static final int VM_ADD = 16;      public static final int VM_SUB = 17;
  public static final int VM_MUL = 18;      public static final int VM_DIVMOD = 19;   public static final int VM_AND = 20;
  public static final int VM_OR = 21;       public static final int VM_XOR = 22;      public static final int VM_SHL = 23;
  public static final int VM_SHR = 24;      public static final int VM_ZERO_EXIT = 25;public static final int VM_INC = 26;
  public static final int VM_DEC = 27;      public static final int VM_IN = 28;       public static final int VM_OUT = 29;
  public static final int VM_WAIT = 30;


 /**
  * Returns the value in the opposite endian
  *
  * @return int
  */
  public int switchEndian(int value)
  {
    int b1 = (value >>  0) & 0xff;
    int b2 = (value >>  8) & 0xff;
    int b3 = (value >> 16) & 0xff;
    int b4 = (value >> 24) & 0xff;
    return b1 << 24 | b2 << 16 | b3 << 8 | b4 << 0;
  }


 /**
  * Load an image file. Will do nothing if the file does
  * not exist.
  */
  public void loadImage(String name)
  {
    int i;
    RandomAccessFile in = null;

    boolean exists = (new File(name)).exists();
    if (exists)
    {
      try
      {
        in = new RandomAccessFile(name, "r");
        i = 0;

        while (i < 5000000)
        {
          memory[i] = in.readInt(); i++;
        }
        if (in != null)
          in.close();
      }
      catch(Exception e) { System.out.println(e); }
    }
    else
    {
      memory[0] = 0;
    }
  }


 /**
  * Save the current image to a file.
  */
  public void saveImage(String name)
  {
    int i;

    try
    {
      RandomAccessFile out = new RandomAccessFile(name, "rw");
      i = 0;

      while (i < 5000000)
      {
        out.writeInt(memory[i]); i++;
      }
      if (out != null)
        out.close();
    }
    catch(Exception e) { System.out.println(e); }
  }


 /**
  * Generic constructor
  */
  public void retro()
  {
  }


 /**
  * Setup and restore the terminal for key breaking and
  * no echoing of the characters being read. This *requires*
  * a Unix-like host. This is one place .NET is better than
  * Java: it at least tries to provide real console access.
  */
  public static void set_tty(boolean on)
  {
    try
    {
      String[] cmd = { "/bin/sh", "-c", "/bin/stty " + (on ? "-echo -icanon min 1" : "echo") + " < /dev/tty" };
      Process p = Runtime.getRuntime().exec(cmd);
      p.waitFor();
    }
    catch (IOException e) { }
    catch (InterruptedException e) { }
  }


 /**
  * Attempt to load the initial image. This tries to load a
  * file (retroImage). If that fails, it uses a built-in one.
  */
  public void initial_image()
  {
    int i;
    for(i = 0; i < 5000000; i++)
      memory[i] = 0;

    loadImage("retroImage");
    if (memory[0] == 0)
    {

File: dotnet/top.cs
/******************************************************
 * Ngaro for Mono / .NET
 *
 * Copyright (c) 2009, Simon Waite and Charles Childers
 *
 * Please compile with `gmcs` as `mcs` seems to have a
 * simple Console.in implementation.
 ******************************************************/

using System;
using System.IO;

namespace Retro.Forth
{
  public class VM
  {
    int sp, rsp, ip;
    int[] data;
    int[] address;
    int[] ports;
    int[] memory;

    enum OpCodes
    {
      VM_NOP,       VM_LIT,       VM_DUP,
      VM_DROP,      VM_SWAP,      VM_PUSH,
      VM_POP,       VM_CALL,      VM_JUMP,
      VM_RETURN,    VM_GT_JUMP,   VM_LT_JUMP,
      VM_NE_JUMP,   VM_EQ_JUMP,   VM_FETCH,
      VM_STORE,     VM_ADD,       VM_SUB,
      VM_MUL,       VM_DIVMOD,    VM_AND,
      VM_OR,        VM_XOR,       VM_SHL,
      VM_SHR,       VM_ZERO_EXIT, VM_INC,
      VM_DEC,       VM_IN,        VM_OUT,
      VM_WAIT
    }


    public VM()
    {
      sp = 0;
      rsp = 0;
      ip = 0;
      data    = new int[100];
      address = new int[100];
      ports   = new int[1024];
      memory  = new int[5000000];
      Retro();
    }


    public int switchEndian(int value)
    {
      int b1 = (value >>  0) & 0xff;
      int b2 = (value >>  8) & 0xff;
      int b3 = (value >> 16) & 0xff;
      int b4 = (value >> 24) & 0xff;
      return b1 << 24 | b2 << 16 | b3 << 8 | b4 << 0;
    }


    public void loadImage()
    {
      int i;
      if (!File.Exists("retroImage"))
        return;

      BinaryReader binReader = new BinaryReader(File.Open("retroImage", FileMode.Open));
      try
      {
        // If the file is not empty,
        // read the application settings.
        // First read 4 bytes into a buffer to
        // determine if the file is empty.
        byte[] testArray = new byte[3];
        int count = binReader.Read(testArray, 0, 3);

        if (count != 0)
        {
          // Reset the position in the stream to zero.
          binReader.BaseStream.Seek(0, SeekOrigin.Begin);

          i = 0;
          while (i < 5000000)
          {
            memory[i] = binReader.ReadInt32(); i++;
          }
        }
      }

      // If the end of the stream is reached before reading
      // the four data values, ignore the error and use the
      // default settings for the remaining values.
      catch(EndOfStreamException e)
      {
        Console.WriteLine("{0} caught and ignored. " + "Using default values.", e.GetType().Name);
      }
      finally
      {
        binReader.Close();
      }
    }


    public void saveImage()
    {
      int i;
      BinaryWriter binWriter = new BinaryWriter(File.Open("retroImage", FileMode.Create));
      try
      {
        i = 0;
        while (i < 5000000)
        {
          binWriter.Write(memory[i]); i++;
        }
      }

      // If the end of the stream is reached before reading
      // the four data values, ignore the error and use the
      // default settings for the remaining values.
      catch(EndOfStreamException e)
      {
        Console.WriteLine("{0} caught and ignored. " + "Using default values.", e.GetType().Name);
      }
      finally
      {
        binWriter.Close();
      }
    }


    public void Retro()
    {
      loadImage();

      if (memory[0] == 0)
      {
// end top